## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [94]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [95]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [96]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [97]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.count()

first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [98]:
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [99]:
df["category"].unique()

array(['decor', 'consumables', 'software', 'furniture', 'electronics'],
      dtype=object)

In [100]:
df["category"].nunique()

5

In [101]:
df["category"].value_counts()

category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64

In [102]:
df["category"].value_counts().sum()

54639

In [103]:
df["unit_cost"].sum()

5433633.970000001

In [104]:
# What three item categories had the most entries?
df["category"].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [105]:
# For the category with the most entries, 
# which subcategory had the most entries?
df.loc[df["category"] == "consumables", "subcategory"].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [106]:
# Which five clients had the most entries in the data?
df["client_id"].value_counts().head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [107]:
# Store the client ids of those top 5 clients in a list.
top_five_clients = list(df["client_id"].value_counts().head(5).index)
top_five_clients

[33615, 66037, 46820, 38378, 24741]

In [108]:
# How many total units (the qty column) did the
# client with the most entries order order?
df.loc[df["client_id"] == 33615, "qty"].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [109]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df["subtotal"] = df["unit_price"] * df["qty"]
df[['unit_price', 'qty', 'subtotal']].head()

unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00

In [110]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def pound_price(weight):
    if weight > 50:
        return weight * 7
    return weight * 10
df["total_weight"] = df["unit_weight"] * df["qty"]
df["shipping_price"] = df["total_weight"].apply(pound_price)
df[["unit_weight", "qty", "total_weight", "shipping_price"]].head(3)

unit_weight  qty  total_weight  shipping_price
0         7.50  105        787.50         5512.50
1         1.49   21         31.29          312.90
2         1.68   39         65.52          458.64

In [111]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
# sales_tax = (df["subtotal"] + df["shipping_price"]) *.0925
df["total_price"] = round((df["subtotal"] + df["shipping_price"]) * 1.0925 , 2)
df[["subtotal", "shipping_price", "total_price"]].head(3)


subtotal  shipping_price  total_price
0  115164.00         5512.50    131839.08
1     523.95          312.90       914.26
2     527.28          458.64      1077.12

In [112]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df["total_cost"] = (df["unit_cost"] * df["qty"]) + df["shipping_price"]
df[["unit_cost", "qty", "shipping_price", "total_cost"]].head(3)


unit_cost  qty  shipping_price  total_cost
0     762.71  105         5512.50    85597.05
1      15.09   21          312.90      629.79
2       7.86   39          458.64      765.18

In [113]:
# Create a column for the profit of
# each line using line cost and line price
df["total_profit"] = round(df["total_price"] - df["total_cost"] , 2)
df[["total_cost", "total_price", "total_profit"]].head(3) 


total_cost  total_price  total_profit
0    85597.05    131839.08      46242.03
1      629.79       914.26        284.47
2      765.18      1077.12        311.94

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [114]:
# Check your work using the totals above
order_ids = [2742071, 2173913, 6128929]

def find_order_price(order_id):
    order_lines_df = df.loc[df["order_id"] == order_id, "total_price"]
    return round(order_lines_df.sum(), 2)

for order_id in order_ids:
    print(f"Order {order_id} total: ${find_order_price(order_id)}")

Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [122]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
def client_totals(client_id, column):
    client_df = df.loc[df["client_id"] == client_id, column]
    return round(client_df.sum(), 2)

for client_id in top_five_clients:
    print(f"{client_id}: ${client_totals(client_id, 'total_price')}")


33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [134]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
data_dict = {"client_id": top_five_clients}
columns = ["qty", "shipping_price", "total_price", "total_cost", "total_profit"] 
for col in columns:
    data_dict[col] = [client_totals(client_id, col) for client_id in top_five_clients]
summary_df = pd.DataFrame(data_dict)
summary_df = summary_df.sort_values("total_profit", ascending=False)
summary_df.head()

client_id     qty  shipping_price  total_price   total_cost  total_profit
4      24741  239862      5126448.37  82268892.02  45688899.71   36579992.31
3      38378   73667      3429455.40  12906550.87   9634720.98    3271829.89
1      66037   43018      1395151.85  10259514.79   7004482.98    3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64    2736602.72
0      33615   64313      1828984.89   8377308.52   6175313.91    2201994.61

In [136]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
summary_df = df.rename(columns={"client_id":"Client ID", 
                                "qty":"Units", 
                                "shipping_price":"Shipping", 
                                "total_price":"Total Revenue", 
                                "total_cost":"Total Cost",
                                "total_profit":"Total Profit"})

def convert_to_millions(df, columns):
    summary_df_final = summary_df.copy()
    for col in columns:
        summary_df_final[col] /= 1000000
    return summary_df_final


summary_df = summary_df.apply(convert_to_millions, axis = 1)
summary_df.head(5)


TypeError: convert_to_millions() missing 1 required positional argument: 'columns'

In [ ]:
# Sort the updated data by "Total Profit" form highest to lowest
summary_df_final = summary_df_final.sort_values("Total Profit", ascending=False)
summary_df_final.head(5)

In [ ]:
# Write a brief 2-3 sentence summary of your findings.